test/library/gitlabidp.go

```diff
@@ -0,0 +1,316 @@
+package library
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	"gopkg.in/square/go-jose.v2/json"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	routev1client "github.com/openshift/client-go/route/clientset/versioned/typed/route/v1"
+
+	"github.com/stretchr/testify/require"
+)
+
+func AddGitlabIDP( // TODO: possibly make this be a wrapper to a function to simplify cleanup
+	t *testing.T,
+	kubeconfig *rest.Config,
+) (idpURL, idpName string, cleanups []func()) {
+	kubeClients, err := kubernetes.NewForConfig(kubeconfig)
+	require.NoError(t, err)
+
+	routeClient, err := routev1client.NewForConfig(kubeconfig)
+	require.NoError(t, err)
+
+	configclients, err := configv1client.NewForConfig(kubeconfig)
+	require.NoError(t, err)
+
+	nsName, gitlabHost, cleanup := deployPod(t, kubeClients, routeClient,
+		"gitlab",
+		"docker.io/gitlab/gitlab-ce:latest",
+		[]corev1.EnvVar{
+			// configure password for GitLab root user
+			{Name: "GITLAB_OMNIBUS_CONFIG", Value: "gitlab_rails['initial_root_password']='password';"},
+		},
+		[]corev1.Volume{
+			{
+				Name: "configdir",
+				VolumeSource: corev1.VolumeSource{
+					EmptyDir: &corev1.EmptyDirVolumeSource{},
+				},
+			},
+		},
+		[]corev1.VolumeMount{
+			{
+				Name:      "configdir",
+				SubPath:   "gitlab/config",
+				MountPath: "/etc/gitlab",
+			},
+			{
+				Name:      "configdir",
+				SubPath:   "gitlab/logs",
+				MountPath: "/var/log/gitlab",
+			},
+			{
+				Name:      "configdir",
+				SubPath:   "gitlab/data",
+				MountPath: "/var/opt/gitlab",
+			},
+		},
+		corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				// it's a bit resource-heavy, it will probably eat more during start-up
+				// max observed cpu: 1420m
+				// max observed memory: 2001Mi
+				// but that's more than our devel clusters provide, let's just request some minimum
+				// and have the pod eat more
+				"cpu":    resource.MustParse("500m"),
+				"memory": resource.MustParse("1500Mi"),
+			},
+		},
+	)
+	cleanups = []func(){cleanup}
+
+	gitlabURL := "https://" + gitlabHost
+	transport, err := rest.TransportFor(kubeconfig)
+	defer func() {
+		if err != nil {
+			for _, c := range cleanups {
+				c()
+			}
+		}
+	}()
+	require.NoError(t, err)
+
+	gitlabClient := GitLabClientFor(t, transport, gitlabURL)
+
+	// TODO: future us might try to configure whitelist IPs so that we can configure
```

Comment:
(No action required) Is there a convention to ensure ownership of TODOs e.g. `TODO(marun)`? I recall this from upstream but maybe this isn't common downstream.

Reply:
I've always thought of TODOs as of useful ideas if you want to improve parts of the code but don't currently have time/willingness to deal with that, and anyone can do that if they find time for it.

If there was a TODO that I would like to own as I really just hacked around something and want to fix that soon, I would go with TODO(owner) as you suggest.